xarray/core/dataarray.py

```diff
@@ -230,6 +235,9 @@ def __init__(self, data, coords=None, dims=None, name=None,
         # uncomment for a useful consistency check:
         # assert all(isinstance(v, Variable) for v in coords.values())
 
+        # check shape consistency
+        _check_shape_consistency(variable.shape, coords, variable.dims)
```

Comment:
This could be pretty expensive — it means we have to check shape consistency every time we create a DataArray, even if we know that the shape is already consistent, e.g., because it was created by another xarray operation.